### PR TITLE
Decode QR codes from Google Authenticator's export function

### PR DESCRIPTION
**Pull request by sssudame**
Add the ability so provide an image file as the infile. If the file contains a QR code generated by Google Authenticator's "Transfer Accounts" function, it will be decoded directly in a single step. This is meant to help users who need to access their secrets from Google Authenticator but don't have a QR code decoder and don't want to use an online one due to security concerns.

### CODE REVIEW

**Comment by scito** [comment]
Hi Shreerang, thanks for your PR. This seems to be really a useful enhancement!
Why did you close this PR?
I'd like to integrate this functionality. Will you resubmit your code?

File: requirements-dev.txt [hunk 1]
```diff
@@ -2,3 +2,5 @@ wheel
 pytest
 flake8
 pylint
+qreader
```
Comment (scito) [inline requirements-dev.txt:5]: Not necessary as it is already in the requierments.txt

File: extract_otp_secret_keys.py [hunk 2]
```diff
@@ -76,12 +82,19 @@ def main(sys_args):
...
+    arg_parser.add_argument('infile',
+                            help="image file containing a QR code from a Google Authenticator export or a text file "
+                                 "or - for stdin with \"otpauth-migration://...\" URLs separated by newlines. Lines "
+                                 "starting with # are ignored.")
     arg_parser.add_argument('--json', '-j', help='export json file or - for stdout', metavar=('FILE'))
     arg_parser.add_argument('--csv', '-c', help='export csv file or - for stdout', metavar=('FILE'))
-    arg_parser.add_argument('--keepass', '-k', help='export totp/hotp csv file(s) for KeePass, - for stdout', metavar=('FILE'))
-    arg_parser.add_argument('--printqr', '-p', help='print QR code(s) as text to the terminal (requires qrcode module)', action='store_true')
-    arg_parser.add_argument('--saveqr', '-s', help='save QR code(s) as images to the given folder (requires qrcode module)', metavar=('DIR'))
+    arg_parser.add_argument('--keepass', '-k', help='export totp/hotp csv file(s) for KeePass, - for stdout',
```
Comment (scito) [inline extract_otp_secret_keys.py:91]: I suggest to revert the reformatting, i.e. the new line.

File: extract_otp_secret_keys.py [hunk 3]
```diff
@@ -97,59 +110,111 @@ def extract_otps(args):
...
+            if not quiet:
+                print()
+
+            otps.append(otp)
+
     return otps
 
 
+def get_lines_from_file(filepath):
+    global verbose
```
Comment (scito) [inline extract_otp_secret_keys.py:157]: args should be passed instead this global variable

File: extract_otp_secret_keys.py [hunk 4]
```diff
@@ -97,59 +110,111 @@ def extract_otps(args):
...
-
-                otps.append(otp)
-    finally:
-        finput.close()
+
+    for line in lines:
+        if verbose:
+            print(line)
+        if line.startswith('#') or line == '':
+            continue
```
Comment (scito) [inline extract_otp_secret_keys.py:121]: This seems not necessary anymore

File: extract_otp_secret_keys.py [hunk 5]
```diff
@@ -47,12 +47,18 @@
...
 from urllib.parse import parse_qs, urlencode, urlparse, quote
 from os import path, makedirs
 from re import compile as rcompile
 import protobuf_generated_python.google_auth_pb2
 
 
+verbose = False
+quiet = True
+
+
```
Comment (scito) [inline extract_otp_secret_keys.py:61]: These variables are not necessary. They are read from arg_parser.

File: Pipfile [hunk 6]
```diff
@@ -4,9 +4,15 @@ verify_ssl = true
...
-protobuf = "*"
+protobuf = "==4.21.12"
 qrcode = "*"
 pillow = "*"
+wheel = "==0.38.4"
+pytest = "==7.2.0"
+flake8 = "==6.0.0"
+pylint = "==2.15.9"
+qreader = "==1.3.1"
+opencv-python = "==4.6.0.66"
```
Comment (scito) [inline Pipfile:15]: I think fixing the version is not necessary.

File: extract_otp_secret_keys.py [hunk 7]
```diff
@@ -97,59 +110,111 @@ def extract_otps(args):
...
 
 
+def get_lines_from_file(filepath):
+    global verbose
+
+    # Check if this is an image file
+    if(path.splitext(filepath)[1][1:].lower() in ('bmp', 'jpg', 'jpeg', 'png', 'tif', 'tiff')):
+        # It's an image file, so try to read it as a QR Code
+        try:
+            decoder = QReader()
```
Comment (scito) [inline extract_otp_secret_keys.py:163]: The decoder creation could be moved down, right before the usage.

File: extract_otp_secret_keys.py [hunk 8]
```diff
@@ -97,59 +110,111 @@ def extract_otps(args):
...
+        finally:
+            finput.close()
+        return lines
+
+
 def get_payload_from_line(line, i, args):
     global verbose
     if not line.startswith('otpauth-migration://'):
-        eprint('\nWARN: line is not a otpauth-migration:// URL\ninput file: {}\nline "{}"\nProbably a wrong file was given'.format(args.infile, line))
+        eprint(
```
Comment (scito) [inline extract_otp_secret_keys.py:204]: Revert the formatting changes

**Review by scito — COMMENTED** [review]
I've had a quick look.

**Comment by scito** [comment]
I'll work on this PR myself and do the refactoring. I really like this functionality. I was not aware of these Python modules.
Thanks for your contribution.

**Comment by sssudame** [comment]
Hey Roland

Thanks for your email, so sorry about the delay in responding, I was
caught up in some personal business.

Did you actually complete the refactoring yourself? Do you need any help
from me?

Thanks for taking this on!

Shreerang

**Comment by scito** [comment]
Hi Shreerang

I've done the refactoring and much more. This Python script can now capture QR codes itself, see https://github.com/scito/extract_otp_secrets/releases/tag/v2.0.0

This work is based on your input!

If you see anything to improve, your welcome to submit a PR.

Thanks again for your contribution!

Best regards,
Roland (scito)

------- Original Message -------
Shreerang Sudame ***@***.***> schrieb am Donnerstag, 19. Januar 2023 um 23:53:

**Comment by sssudame** [comment]
Great, thank you!

**Pull request closed without merging**